File: java/Kode/Weapon.java
import java.util.UUID;

// Kelas dibuat Abstract karena Weapon dasar tidak seharusnya diinstansiasi
public abstract class Weapon {
    private String name;
    private int level;
    private String tier;
    private WeaponMaterial material; // Composition
    private UUID id;

    public Weapon(String name, int level, String tier, WeaponMaterial material) {
        this.name = name;
        this.level = level;
        this.tier = tier;
        this.material = material; // Composition
        this.id = UUID.randomUUID();
    }
    
    public String getName() {
        return name;
    }

    public int getLevel() {
        return level;
    }

    // Method abstract yang wajib di-implementasi (override) oleh Subclass
    protected abstract String specificsInfo();

    // Method polimorfik utama
    public void displayWeaponDetails(boolean fullDetail) {
        System.out.println("  [WEAPON] Name: " + name + " (Lvl " + level + ", Tier: " + tier + ")");
        
        if (fullDetail && material != null) {
            System.out.println("  --- Material (Composition) ---");
            material.displayMaterial();
        }
        
        String info = specificsInfo();
        if (info != null && !info.isEmpty()) {
            System.out.println("  Specifics: " + info);
        }
    }
}

File: cpp/Kode/Weapon.cpp
#include "Weapon.h"
#include "WeaponMaterial.h" 
#include <utility>

Weapon::Weapon(std::string n, int l, std::string t, const std::vector<std::pair<std::string, double>>& s,
               std::unique_ptr<WeaponMaterial> m) : 
    Name(n), Level(l), Tier(t), stats(s), material(std::move(m)) {}

void Weapon::displayWeapon() const {
    // Dipanggil saat Print Weapon (LENGKAP)
    std::cout << "  [WEAPON] Name: " << Name << " (Lvl " << Level << ", Tier " << Tier << ")" << std::endl;
    std::cout << "  Stats: ";
    for (const auto& stat : stats) {
        std::cout << stat.first << ": " << stat.second << "; ";
    }
    std::cout << std::endl;
    
    // Tampilkan Material (Komposisi)
    std::cout << "  --- Material (Komposisi) ---" << std::endl;
    if (material) {
        material->displayMaterial();
        std::cout << std::endl;
    } else {
        std::cout << "  No material defined." << std::endl;
    }
}

void Weapon::displayWeaponWithoutMaterial() const {
    // Dipanggil saat Print Character (HANYA Weapon)
    std::cout << "  [WEAPON] Name: " << Name << " (Lvl " << Level << ", Tier " << Tier << ")" << std::endl;
    std::cout << "  Stats: ";
    for (const auto& stat : stats) {
        std::cout << stat.first << ": " << stat.second << "; ";
    }
    std::cout << std::endl;
}